File: Algorithms/1171.Remove-Zero-Sum-Consecutive-Nodes-from-Linked-List/remove_zero_sum_consecutive_nodes_from_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeZeroSumSublists(self, head: ListNode) -> ListNode:
        p = head
        records = {}
        pre_sum = 0
        while p:
            pre_sum += p.val
            if pre_sum == 0:
                head = p.next
                records.clear()
                p = p.next
            else:
                if pre_sum in records:
                    records[pre_sum].next = p.next
                    p = head
                    pre_sum = 0
                    records.clear()
                else:
                    records[pre_sum] = p
                    p = p.next
        return head
                

File: Algorithms/1171.Remove-Zero-Sum-Consecutive-Nodes-from-Linked-List/remove_zero_sum_consecutive_nodes_from_linked_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeZeroSumSublists(head *ListNode) *ListNode {
    records := make(map[int]ListNode)
    p := head
    pre_sum := 0
    for {
        if p == nil {
            break
        }
        pre_sum += p.Val
        // 总数为0 则前面的都可弃掉，直接从next开始
        if pre_sum == 0 {
            head = p.Next
        } else {
            records[pre_sum] = *p
        }
        p = p.Next
    }
    
    p = head
    pre_sum = 0
    for {
        if p == nil {
            break
        }
        pre_sum += p.Val
        p.Next = records[pre_sum].Next
        p = p.Next
    }
    return head
}
